src/memory_region/mod.rs

```diff
@@ -0,0 +1,181 @@
+/// Local Memory Region
+mod local;
+/// Raw Memory Region
+mod raw;
+/// Remote Memory Region
+mod remote;
+
+#[allow(unreachable_pub)]
+pub use local::{LocalMr, LocalMrAccess, LocalMrSlice};
+pub(crate) use raw::RawMemoryRegion;
+#[allow(unreachable_pub)]
+pub use remote::{RemoteMr, RemoteMrAccess, RemoteMrSlice};
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use utilities::cast_to_ptr;
+
+/// Rdma Memory Region Access
+pub trait MrAccess: Sync + Send + Debug {
+    /// Get the start addr
+    fn addr(&self) -> usize;
+
+    /// Get the length
+    fn length(&self) -> usize;
+
+    /// Get the remote key
+    fn rkey(&self) -> u32;
+
+    /// Get the token
+    #[inline]
+    fn token(&self) -> MrToken {
+        MrToken {
+            addr: self.addr(),
+            len: self.length(),
+            rkey: self.rkey(),
+        }
+    }
+}
+
+/// Memory region token used for the remote access
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Clone, Copy, Debug)]
+pub struct MrToken {
+    /// The start address
+    pub addr: usize,
+    /// The length
+    pub len: usize,
+    /// The rkey
+    pub rkey: u32,
+}
+
+/// Mr slice metadata, which size should equal to sizeof usize
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct SliceMetaData {
+    /// start at the offset of Mr which this slice belong
+    offset: u32,
+    /// len of mr slice
+    len: u32,
+}
+
+impl From<usize> for SliceMetaData {
+    fn from(u: usize) -> Self {
+        unsafe { *(cast_to_ptr(&u)) }
+    }
+}
+
+impl From<SliceMetaData> for usize {
+    #[inline]
+    fn from(s: SliceMetaData) -> Self {
+        unsafe { *(cast_to_ptr(&s)) }
+    }
+}
+
+/// Slice index trait, copy of unstable rust std lib
```

Comment:
It's unstable in the rust std lib for some reasons, I don't recommend to simply copy it here.